### DISCUSSION TITLE
v1.4.0 - Demo site / Build Errors

### DISCUSSION THREAD

**Discussion by JeepNL** [microsoft/fluentui-blazor#184] [General]
1. There isn't a demo site running anymore? See https://github.com/microsoft/fast-blazor/pull/97 and (PR) https://github.com/microsoft/fast-blazor/pull/142
. 
2. I'm trying to build the Fast-Blazor WASM examples but I'm getting build errors (ASP.NET Core 6.0.4 and 7 prev. 3) on the `fast-blazor-main\examples\FluentUI.Demo.Shared` project

```
Microsoft (R) Build Engine version 17.2.0-preview-22175-02+058a0262c for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored C:\Users\jdkra\Downloads\fast-blazor-main\examples\FluentUI.Demo.Shared\FluentUI.Demo.Shared.csproj (in 280 ms).
  You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  The given key 'Common.GloballyQualifiedTypeName' was not present in the dictionary.
     at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
     at Microsoft.AspNetCore.Razor.Language.Components.ComponentGenericTypePass.Visitor.RewriteTypeNames(TypeNameRewriter rewriter, ComponentIntermediateNode node, Nullable`1 hasT
  ypeArgumentSpecified, IDictionary`2 bindings)
     at Microsoft.AspNetCore.Razor.Language.Components.ComponentGenericTypePass.Visitor.Process(ComponentIntermediateNode node)
     at Microsoft.AspNetCore.Razor.Language.Components.ComponentGenericTypePass.Visitor.VisitComponent(ComponentIntermediateNode node)
     at Microsoft.AspNetCore.Razor.Language.Intermediate.IntermediateNodeWalker.VisitDefault(IntermediateNode node)
     at Microsoft.AspNetCore.Razor.Language.Intermediate.IntermediateNodeWalker.VisitDefault(IntermediateNode node)
     at Microsoft.AspNetCore.Razor.Language.Intermediate.IntermediateNodeWalker.VisitDefault(IntermediateNode node)
     at Microsoft.AspNetCore.Razor.Language.Intermediate.IntermediateNodeWalker.VisitDefault(IntermediateNode node)
     at Microsoft.AspNetCore.Razor.Language.Intermediate.IntermediateNodeWalker.VisitDefault(IntermediateNode node)
     at Microsoft.AspNetCore.Razor.Language.Intermediate.IntermediateNodeWalker.VisitDefault(IntermediateNode node)
     at Microsoft.AspNetCore.Razor.Language.Intermediate.IntermediateNodeWalker.VisitDefault(IntermediateNode node)
     at Microsoft.AspNetCore.Razor.Language.Intermediate.IntermediateNodeWalker.VisitDefault(IntermediateNode node)
     at Microsoft.AspNetCore.Razor.Language.Intermediate.IntermediateNodeWalker.VisitDefault(IntermediateNode node)
     at Microsoft.AspNetCore.Razor.Language.DefaultRazorOptimizationPhase.ExecuteCore(RazorCodeDocument codeDocument)
     at Microsoft.AspNetCore.Razor.Language.DefaultRazorEngine.Process(RazorCodeDocument document)
     at Microsoft.AspNetCore.Razor.Language.RazorProjectEngine.Process(RazorProjectItem projectItem)
     at Microsoft.NET.Sdk.Razor.Tool.GenerateCommand.<>c__DisplayClass55_0.<GenerateCode>b__0(Int32 i)
     at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
  --- End of stack trace from previous location ---
     at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
     at System.Threading.Tasks.TaskReplicator.Replica.Execute()
C:\Program Files\dotnet\sdk\7.0.100-preview.3.22179.4\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.CodeGeneration.targets(150,5): error : rzc generate exited with
code 1. [C:\Users\jdkra\Downloads\fast-blazor-main\examples\FluentUI.Demo.Shared\FluentUI.Demo.Shared.csproj]

```

**Comment by vnbaaij** [comment]
1) The demo is still there. You can find it at https://brave-cliff-0c0c93310.azurestaticapps.net/. The one you were referring to is part of the GitHub workflow where an Azure Static Web app is setup automatically to test a PR (hence the 142 in the URL, the number of the PR). It gets deleted once the PR is closed/merged.

2) We have not done any testing yet with .NET 7. I'll try and have a look next week.

**Comment by JeepNL** [comment]
@vnbaaij Thank you for quick reply, even if it was on the weekend! (EDIT: especially because it was on the weekend! 😉)

About the online demo: I couldn't find a link anywhere here, it's not in the `README.md`, the only reference to it is in your answer here, see: https://github.com/microsoft/fast-blazor/search?q=https%3A%2F%2Fbrave-cliff-0c0c93310.azurestaticapps.net%2F&type=issues. 

The PR link (142) I referred to was the second link I mentioned because I know when the PR was still active it worked and I could see the [Icon Page](https://brave-cliff-0c0c93310.azurestaticapps.net/IconPage) which you mentioned in [this tweet](https://twitter.com/vnbaaij/status/1516446076429127687)

I spent my weekend trying out Fast-Blazor (_trying new things out is what I love to do, so it was a weekend well spent, I learned a couple of new things_) and I ran into several problems with Fast-Blazor Components.

**TL;DR;** I don't think I can use it at the moment, there are too many things I don't understand, can't figure out.

**FYI (1)**: At [Using Design Tokens from C# code](https://github.com/microsoft/fast-blazor#option-1-using-design-tokens-from-c-code) (_preferable option_) **ref1** & **ref3** are `FluentAnchor` and  **ref2** and **ref4** are `FluentButton` but in the code above that there's only `<FluentButton ...`,

**FYI (2)** In the [Demo Website](https://brave-cliff-0c0c93310.azurestaticapps.net/CardPage) at menu option 'Card' when you first load it,  look at the fill color of the second card (_Dark, second from top_). Now switch the 'Light Dark Switch' **twice**, and see what happens. This is something I ran into quite a lot in my code trying to use different colors. This happens also at Card 4 (_Tinted neutral, dark_). If you look at Card 5 (_Custom card background color_) and 6 (_Accent and neutral color by DSP_) the text color in the fist/top button is almost unreadable. I tried to change to text color in dark mode, but I couldn't figure out how to.

**FYI (3)** Clicking on the Demo Website **[Number Field](https://brave-cliff-0c0c93310.azurestaticapps.net/NumberFieldPage)** menu option gives an error.

When I start a new Blazor WASM project there are two things (_actually 3_), important to me: Mobile First, and a Light/Dark theme (_the 3rd is privacy friendly_). So that's where I start when using a new component library.

I'd like to adjust some colors and searching the Fast documentation I found the  [Fast Color Explorer](https://color.fast.design/), it's a great interactive page, and I think it looks really nice, so I tried to use different base colors than the default white/dark grey colors. But I didn't succeed, I think I need more info on how to use not only `AccentBaseColor` but also on (_injecting, preferred code option 1_) `NeutralBaseColor` & `FillColor` and maybe more.

I also ran into [this issue](https://github.com/microsoft/fast-blazor/issues/125) about CSS Isolation and although I could made it to work, some things aren't clear to me.

I know I mention 'things that do not work as expected' and 'are not clear' to me a lot, and I could share my code I used in my project, but to make it all clear I think I need to spend a couple of hours extra on this issue. I'm Dutch, so I need more time to express myself in English.

I really would like to use Fast-Blazor but for me to be able to use it I need more sample code. I don't need it now, or tomorrow (_or next month_), I know you all are very busy and working hard on Fast-Blazor Components and I think the promise is great and it has a lot of potential. I also really like that is pretty small in size compared to other Blazor component libraries. I'm developing Blazor WASM apps only, so size does matter to me: as small as possible :).

**Comment by JeepNL** [comment]
Maybe this 'issue' should be moved to Discussions?

**Comment by vnbaaij** [comment]
Good plan. I will do that

**Comment by vnbaaij** [comment]
The reason we are not publishing/promoting the demo site is we don't feel it is ready yet. Yes, it will show you basic usage, but it does not have code samples etc. Once we are done, it will be made available with a more friendly URL.

On FYI (1): The `FluentAnchor` references were a leftover for earlier. There should only be `FluentButton` references. I've corrected this in the README

**Reply by JeepNL** [reply]
Ah, okay, I see. That makes sense (demo site)

**Comment by JeepNL** [comment]
### About Blazor CSS Isolation. 

This is just FYI and it's possible that 'm doing something completely wrong. I just stumbled on this and I couldn't figure out why

With `FluentCard` I needed to put some properties in the `style=` argument and use an extra `<div>` inside the component to display it as I wanted. I expected I could just put them all in the reference `class=footercontent`, but that doesn't work.

Blazor *Component*. Access it from any `.razor` page with `<Footer />`

**Footer.razor**
```
<p>Paragraph just for testing 'margin-top' below</p>

<FluentCard class="footercontent" style="margin-top:2rem;margin-bottom:1rem;max-width:720px;">
    <div>
        Footer Component Content with <a href="">href</a>
    </div>
</FluentCard>

<p>Paragraph just for testing 'margin-bottom' above</p>
```

**Footer.razor.css**
```
.footercontent {
}

.footercontent div {
    padding: .5rem;
    font-family: monospace;
    font-size: 75%;
    line-height: 150%;
	color: dimgray;
}

.footercontent a {
    color: orangered;
}

    .footercontent a:hover {
        color: orange;
    }

```

### I expected it to work as below:
(without the `div` element)

**Footer.razor**
```
<p>Paragraph just for testing 'margin-top' below</p>

<FluentCard class="footercontent">
        Footer Component Content with <a href="">href</a>
</FluentCard>

<p>Paragraph just for testing 'margin-bottom' above</p>
```

**Footer.razor.css**
```
.footercontent {
    margin-top: 2rem;
    margin-bottom: 1rem
    ;max-width: 720px;
    padding: .5rem;
    font-family: monospace;
    font-size: 75%;
    line-height: 150%;
    color: dimgray;
}

.footercontent a {
    color: orangered;
}

    .footercontent a:hover {
        color: orange;
    }

```

**Reply by vnbaaij** [reply]
Did you try to use the `::deep` pseudo-element? (see https://docs.microsoft.com/en-us/aspnet/core/blazor/components/css-isolation)

**Reply by JeepNL** [reply]
Yes, I did try `::deep` as well at several places (beginning / EOL), but If I understand correctly the pseudo-element is for child components, in this case child components of `Footer.razor`, and it has none.

**Reply by JeepNL** [reply]
fyi: I've read this issue too (css isolation): https://github.com/microsoft/fast-blazor/issues/125

**Reply by JeepNL** [reply]
The strange thing with the example above: look at the top: `.footercontent { }` is empty, anything I put there has no effect.  But at the bottom: 

```
.footercontent a {
    color: orangered;
}

    .footercontent a:hover {
        color: orange;
    }
```

does work on the `<FluentCard class="footercontent" ...>` element itself (without the `<div>` inside it).

**Comment by vnbaaij** [comment]
FYI (3) is fixed now!
